### ISSUE TITLE
Golang runtime doesn't support fully go.mod 

### ISSUE THREAD

**Issue by 1ricardo** [kubeless/runtimes#84]
Hello,

I notice that **Kubless** doesn't support replace sections in the go.mod file.

 For example, even if I create the following go.mod file:
```
module gitbranch

go 1.14

require (
	github.com/go-git/go-git/v5 v5.1.0
	github.com/kubeless/kubeless v1.0.7
	github.com/logrusorgru/aurora/v3 v3.0.0
)

replace github.com/logrusorgru/aurora/v3 v3.0.0 => ../localtest                                            
 ```

from the above example, you can notice that in **replace** section, I overwrite the packages aurora with my local test branch,
in this scenario, the compilation should **fail**, but as you see form below output, it's working successfully using the github repo.


Kubeless output:

```
gitbranch-get   kbox            gitbranch.Get           go1.14  module gitbranch                                        1/1 READY    
                                                                                                                                     
                                                                go 1.14                                                              
                                                                                                                                     
                                                                require (                                                            
                                                                        github.com/go-git/go-git/v5 v5.1.0                                   
                                                                        github.com/kubeless/kubeless v1.0.7                                  
                                                                        github.com/logrusorgru/aurora/v3 v3.0.0                              
                                                                )                                                                    
                                                                                                                                     
                                                                replace gitlab.com/project_falcon/kubeless/payload                   
                                                                v0.2.4 => ../payload     
```

For me, it's a BIG problem, because currently, I am working on an application that will communicate with Kubernetes API.
Unfortunately, Kubernetes GO LIBs have broken dependency and you have to use the replace section.

example of Kubernetes:

```
module buildimage

go 1.14

require (
	github.com/kubeless/kubeless v1.0.7
	github.com/logrusorgru/aurora/v3 v3.0.0
	gitlab.com/project_falcon/kubeless/payload v0.2.4
	gitlab.com/project_falcon/kubeless/poddefinition/kaniko v0.1.0
	k8s.io/api v0.0.0
	k8s.io/apiextensions-apiserver v0.0.0
	k8s.io/apiserver v0.0.0
	k8s.io/cli-runtime v0.0.0
	k8s.io/client-go v7.0.0+incompatible
	k8s.io/cloud-provider v0.0.0
	k8s.io/cluster-bootstrap v0.0.0
	k8s.io/code-generator v0.0.0
	k8s.io/component-base v0.0.0
	k8s.io/cri-api v0.0.0
	k8s.io/csi-translation-lib v0.0.0
	k8s.io/kube-aggregator v0.0.0
	k8s.io/kube-controller-manager v0.0.0
	k8s.io/kube-proxy v0.0.0
	k8s.io/kube-scheduler v0.0.0
	k8s.io/kubectl v0.0.0
	k8s.io/kubelet v0.0.0
	k8s.io/kubernetes v0.0.0
	k8s.io/legacy-cloud-providers v0.0.0
	k8s.io/metrics v0.0.0
	k8s.io/sample-apiserver v0.0.0
)

replace (
	k8s.io/api => k8s.io/api v0.16.12
	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.16.12
	k8s.io/apimachinery => k8s.io/apimachinery v0.16.12
	k8s.io/apiserver => k8s.io/apiserver v0.16.12
	k8s.io/cli-runtime => k8s.io/cli-runtime v0.16.12
	k8s.io/client-go => k8s.io/client-go v0.16.12
	k8s.io/cloud-provider => k8s.io/cloud-provider v0.16.12
	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.16.12
	k8s.io/code-generator => k8s.io/code-generator v0.16.12
	k8s.io/component-base => k8s.io/component-base v0.16.12
	k8s.io/cri-api => k8s.io/cri-api v0.16.12
	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.16.12
	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.16.12
	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.16.12
	k8s.io/kube-proxy => k8s.io/kube-proxy v0.16.12
	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.16.12
	k8s.io/kubectl => k8s.io/kubectl v0.16.12
	k8s.io/kubelet => k8s.io/kubelet v0.16.12
	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.16.12
	k8s.io/metrics => k8s.io/metrics v0.16.12
	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.16.12
)
```

My application working fine with that go.mod file, when I run it from a local machine, but Kubeless cannot download dependencies, because it seems that it doesn't read the **replace** section at all. 

Kubeless error

```
kubectl logs -n kbox --follow buildimage-79f575949d-bqphk -c compile
go: finding github.com/kubeless/kubeless 2644ad5
go: downloading github.com/kubeless/kubeless v1.0.7-0.20200406085802-2644ad500db0
go: extracting github.com/kubeless/kubeless v1.0.7-0.20200406085802-2644ad500db0
go: downloading golang.org/x/net v0.0.0-20190125091013-d26f9f9a57f3
go: downloading github.com/prometheus/client_golang v0.9.1
go: extracting github.com/prometheus/client_golang v0.9.1
go: downloading github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: downloading github.com/prometheus/common v0.2.0
go: downloading github.com/prometheus/procfs v0.0.0-20190129233650-316cf8ccfec5
go: downloading github.com/golang/protobuf v1.2.0
go: downloading github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973
go: extracting github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973
go: extracting github.com/prometheus/procfs v0.0.0-20190129233650-316cf8ccfec5
go: extracting github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: extracting github.com/golang/protobuf v1.2.0
go: extracting golang.org/x/net v0.0.0-20190125091013-d26f9f9a57f3
go: extracting github.com/prometheus/common v0.2.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: extracting github.com/matttproud/golang_protobuf_extensions v1.0.1
go: kubeless@v0.0.0-00010101000000-000000000000 requires
        k8s.io/api@v0.0.0: reading k8s.io/api/go.mod at revision v0.0.0: unknown revision v0.0.0


```

I will be glad if you can point my mistakes and I can move forward with my project, currently, I'm stuck on Kubeless compilation

**Comment by andresmgot** [comment]
Hi @1ricardo,

It seems that you are talking about two different issues, first this one:

```
replace github.com/logrusorgru/aurora/v3 v3.0.0 => ../localtest                                            
```

This won't work because Kubeless won't include any external folder in your function container. This means that `localtest` won't be available there (and I assume go.mod falls back to the github repository). You can try to zip a folder with that dependency but you will need to include it as a subpackage of your function (or upload the package to github for example).

Then this one:

```
go: kubeless@v0.0.0-00010101000000-000000000000 requires
        k8s.io/api@v0.0.0: reading k8s.io/api/go.mod at revision v0.0.0: unknown revision v0.0.0
```

It's not clear to me that's the result of a wrong `replace`. What makes you think that? (I may be missing something). What I would check is if anything in your code has a dependency with that `kubeless` version (which seems wrong).

**Comment by 1ricardo** [comment]
Hi @andresmgot 
Thank you for your answer, I really appreciate that. 

 to be honest, I don't have big experience with go.mod, because as DevOps I code small pieces of code.
I have 12 different Kubeless functions which are working just fine, but this is my nightmare. 

With your suggestion, I was trying  to fix dependency and I spent another few hours with small progress  

I looked on this [file](https://github.com/kubeless/kubeless/blob/v1.0.7/go.mod#L48) and used following dependency, which should not conflict with Kubeless

```
module buildimage

go 1.14

require (
	github.com/kubeless/kubeless v1.0.7
	github.com/logrusorgru/aurora/v3 v3.0.0
	github.com/peterbourgon/diskv v2.0.1+incompatible // indirect
	k8s.io/api v0.0.0-20180308224125-73d903622b73
	k8s.io/apiextensions-apiserver v0.0.0-20180327033742-750feebe2038
	k8s.io/apimachinery v0.0.0-20180228050457-302974c03f7e
	k8s.io/client-go v7.0.0+incompatible

)
```

Kubeless and build on my local machine trow the same error: 

```
...
go: downloading github.com/modern-go/reflect2 v1.0.1
go: downloading gopkg.in/yaml.v2 v2.2.8
go: downloading golang.org/x/sys v0.0.0-20190813064441-fde4db37ae7a
/go/pkg/mod/k8s.io/client-go@v7.0.0+incompatible/kubernetes/scheme/register.go:22:2: module k8s.io/api@latest found (v0.18.6), but does not contain package k8s.io/api/admissionregistration/v1alpha1

```
I checked a few Github issues on how to fix that and followed go.mod file allows me to compile successfully code on my machine, but Kubeless stile throwing the same issue with a missing package ` module k8s.io/api@latest found (v0.18.6), but does not contain package k8s.io/api/admissionregistration/v1alpha1`.

For me whole example still points to the issue with section `replace` on Kubeless side, I don't know how it is possible

```
module buildimage

go 1.14

require (
	github.com/kubeless/kubeless v1.0.7
	github.com/logrusorgru/aurora/v3 v3.0.0
	github.com/peterbourgon/diskv v2.0.1+incompatible // indirect
	k8s.io/api v0.0.0-20180308224125-73d903622b73
	k8s.io/apiextensions-apiserver v0.0.0-20180327033742-750feebe2038
	k8s.io/apimachinery v0.0.0-20180228050457-302974c03f7e
	k8s.io/client-go v7.0.0+incompatible

)

replace (
	k8s.io/api => k8s.io/api v0.0.0-20180308224125-73d903622b73
	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.0.0-20180327033742-750feebe2038
	k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20180228050457-302974c03f7e
	k8s.io/client-go => k8s.io/client-go v7.0.0+incompatible
)
```

If will be glad to hear any TIP or suggestion. 

br

**Comment by andresmgot** [comment]
Sorry, I am not able to reproduce the issue, I tried that last go.mod but the function got successfully deployed, the only thing I changed was `module buildimage` to `module function` but I used the example `hellowithdeps.go` present in this repository so it's not using those dependencies.

What you can do is modify the deployment with your function and change the script that compiles the function with a loop command:

```patch
      - args:
-        - /compile-function.sh
+        - tail -f /dev/null
        command:
        - sh
        - -c
        env:
        - name: KUBELESS_INSTALL_VOLUME
          value: /kubeless
        - name: KUBELESS_FUNC_NAME
          value: Hello
        - name: GOCACHE
          value: $(KUBELESS_INSTALL_VOLUME)/.cache
```

That would allow you to `kubectl exec` within the container and execute the command manually, comparing the result with the one you get locally.

**Comment by 1ricardo** [comment]
Hi @andresmgot 

I would like to thank you for your support and time, your suggestion always allows me to take a step forward. 

According to your suggestion I jump on Kubeless pod and play around with compilation according to this [script](https://github.com/kubeless/runtimes/blob/master/stable/golang/compile-function.sh). 

I notice that I could successfully compile just my code from path `/kubeless` as standalone packages, but ofc the compilation on this stage is useless without the second part of code which is in path `/server/kubeless`

So I followed the steps from the [script](https://github.com/kubeless/runtimes/blob/master/stable/golang/compile-function.sh) and unfortunately, I got the same error during the attempt of compilation in path `/server/kubeless`

```
/go/pkg/mod/k8s.io/client-go@v7.0.0+incompatible/kubernetes/scheme/register.go:22:2: module k8s.io/api@latest found (v0.18.6), but does not contain package k8s.io/api/admissionregistration/v1alpha1
```

On this step, I don't understand why it happening, if the Golang doesn't read properly the go.mod file from my function or this is the issue of some hidden dependencies. I don' know ...

but I modified the go.mod file in path `/server` with the following example, and I could compile the whole code successfully. 


```diff
module github.com/kubeless/runtimes/golang

go 1.14

replace kubeless => ./function
+ replace k8s.io/api => k8s.io/api v0.0.0-20180308224125-73d903622b73
+ replace k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20180228050457-302974c03f7e

require (
        github.com/kubeless/kubeless v1.0.7
        kubeless v0.0.0-00010101000000-000000000000
)

```

currently, I don't want to rewrite my code on Python, so I thinking about modifying Deployment and overwrite this file by configmap,
 I know this not the ideal solution but at least it will be working
